### PR TITLE
Vessel owns parts

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: ksp_plugin/part.cpp [hunk 1]
```diff
@@ -11,11 +11,20 @@ namespace internal_part {
...
-Part::Part(PartId const part_id, Mass const& mass)
+Part::Part(PartId const part_id,
+           Mass const& mass,
+           std::function<void()> deletion_callback)
     : part_id_(part_id),
       mass_(mass),
       tail_(make_not_null_unique<DiscreteTrajectory<Barycentric>>()),
-      subset_node_(make_not_null_unique<Subset<Part>::Node>()) {}
+      subset_node_(make_not_null_unique<Subset<Part>::Node>()),
+      deletion_callback_(deletion_callback) {}
```
Comment (pleroy) [inline ksp_plugin/part.cpp]: std::move

File: ksp_plugin/vessel.cpp [hunk 2]
```diff
@@ -49,12 +50,52 @@ void Vessel::set_parent(not_null<Celestial const*> const parent) {
   parent_ = parent;
 }
 
-void Vessel::clear_parts() {
-  parts_.clear();
+not_null<Part*> Vessel::InitializeUnloaded(
+    DegreesOfFreedom<Bubble> initial_state) {
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: const&

File: ksp_plugin/vessel.cpp [hunk 3]
```diff
@@ -49,12 +50,52 @@ void Vessel::set_parent(not_null<Celestial const*> const parent) {
...
+  CHECK(parts_.empty());
+  PartId const id = std::numeric_limits<PartId>::max();
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: Make ...max... a static constant next to the declaration of PartId.

File: ksp_plugin/vessel.cpp [hunk 4]
```diff
@@ -49,12 +50,52 @@ void Vessel::set_parent(not_null<Celestial const*> const parent) {
...
+                                 /*deletion_callback=*/nullptr);
+  dummy_part->set_degrees_of_freedom(initial_state);
+  not_null<Part*> result = dummy_part.get();
+  parts_.emplace(id, std::move(dummy_part));
+  return result;
 }
 
-void Vessel::add_part(not_null<Part*> part) {
-  parts_.push_back(part);
+void Vessel::add_part(not_null<std::unique_ptr<Part>> part) {
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: Should be AddPart now.

File: ksp_plugin/vessel.cpp [hunk 5]
```diff
@@ -49,12 +50,52 @@ void Vessel::set_parent(not_null<Celestial const*> const parent) {
...
+  PartId const id = part->part_id();
+  CHECK_NE(id, std::numeric_limits<PartId>::max());
+  parts_.emplace(id, std::move(part));
+  kept_parts_.insert(part.get());
+}
+
+not_null<std::unique_ptr<Part>> Vessel::extract_part(PartId id) {
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: ExtractPart, const

File: ksp_plugin/vessel.cpp [hunk 6]
```diff
@@ -49,12 +50,52 @@ void Vessel::set_parent(not_null<Celestial const*> const parent) {
...
+  CHECK_NE(id, std::numeric_limits<PartId>::max());
+  auto const it = parts_.find(id);
+  CHECK(it != parts_.end());
+  auto result = std::move(it->second);
+  parts_.erase(it);
+  return result;
+}
+
+void Vessel::keep_part(PartId id) {
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: KeepPart, const.

File: ksp_plugin/vessel.cpp [hunk 7]
```diff
@@ -49,12 +50,52 @@ void Vessel::set_parent(not_null<Celestial const*> const parent) {
...
+  CHECK_NE(id, std::numeric_limits<PartId>::max());
+  kept_parts_.insert(FindOrDie(parts_, id).get());
+}
+
+void Vessel::free_parts() {
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: FreeParts.

File: ksp_plugin/vessel.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: In an unnamed namespace, we use it in other functions too.

Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: Oh, you mean the type PartId. Done.

File: ksp_plugin/vessel.cpp (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: Done.

File: ksp_plugin/vessel.cpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: Done.

File: ksp_plugin/vessel.cpp (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: Done.

File: ksp_plugin/vessel.cpp (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: Done.

File: ksp_plugin/part.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/part.cpp]: Done.

File: ksp_plugin/vessel.cpp (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: Done.

**Comment by eggrobin** [comment]
Also switched from adding the dummy part to the vessel's map to putting it in a separate nullable unique_ptr, since nothing guarantees that the part ID won't be our dummy value (and it's only 32-bit). It's also a bit more readable.

File: ksp_plugin/part.hpp [hunk 8]
```diff
@@ -37,6 +38,9 @@ using quantities::Mass;
 // Corresponds to KSP's |Part.flightID|, *not* to |Part.uid|.  C#'s |uint|
 // corresponds to |uint32_t|.
 using PartId = std::uint32_t;
+// An ID given to dummy parts used by vessels that appear unloaded.  Note that
+// nothing prevents an actual part from having this ID.
+constexpr PartId DummyPartId = std::numeric_limits<PartId>::max();
```
Comment (pleroy) [inline ksp_plugin/part.hpp]: Don't we use snake_case for static variables?

File: ksp_plugin/vessel.hpp [hunk 9]
```diff
@@ -135,6 +135,7 @@ class Vessel {
   not_null<Celestial const*> parent_;
   not_null<Ephemeris<Barycentric>*> const ephemeris_;
 
+  std::unique_ptr<Part> dummy_part_;
```
Comment (pleroy) [inline ksp_plugin/vessel.hpp]: I don't quite understand this change, the previous code was nicer, with all the parts owned by the map including the dummy one.  Now we have to manage this special case in addition to the normal one.  What gives?

**Pull request merged**